Translate Chapter 11 to PTBR (#22)
JuliaIntro/JuliaIntroBR.jl: book/src/chap11.asciidoc

[[chap11]]
== Dicionários

> Este capítulo apresenta outra estrutura de dados embutida chamada de dicionário.

Reviewer: ```suggestion
Este capítulo apresenta outra estrutura de dados embutida chamado dicionário.
```

> === Um Dicionário é um Mapeamento

Reviewer: A numeração das linhas está diferente.

> Um _dicionário_ é como um array, mas mais geral. Em um array, os indíces tem que ser inteiros; em um dicionário eles podem ser (quase) de qualquer tipo.

Reviewer: ```suggestion
Um _dicionário_ é como um array, mas mais geral. Em um array, os índices devem ser inteiros e em um dicionário, eles podem (quase) ser de qualquer tipo.
```

Reviewer: O único erro que eu encontrei foi o acento na palavra "índices", de resto é sugestão de minha opinião para dar mais fluidez na leitura.

(((dicionário)))

> Um dicionário contém uma coleção de indíces, que são chamados de _chaves_, e uma coleção de valores. Cada chave é associada a um único valor. A associação de uma chave e um valor é chamada de _par chave-valor_ ou, as vezes, um item.

Reviewer: ```suggestion
Um dicionário contém uma coleção de índices, que são chamados de _chaves_, e uma coleção de valores. Cada chave é associada a um único valor. A associação de uma chave e um valor é chamada de _par chave-valor_ ou, às vezes, de item.
```

(((chave)))(((valor)))(((par chave-valor)))(((item)))

> Em linguagem matemática, um dicionário representa um _mapeamento_ de chave para valores, então você também pode dizer que cada chave “é mapeado” para um valor. Como um exemplo, iremos construir um dicionário que mapeia palavras em Português para palavras em Espanhol, então as chaves e os valores são todos strings.

Reviewer: ```suggestion
Na linguagem matemática, um dicionário representa um _mapeamento_ da chave para os valores, então você também pode dizer que cada chave “é mapeado” para um valor. Como um exemplo, iremos construir um dicionário que mapeia palavras em Português para palavras em Espanhol, com as chaves e os valores todos strings.
```

(((mapeamento)))

A função +Dict+ cria um novo dicionário sem items. Como +Dict+ é um nome de uma função embutida, devemos evitar usar ele como um nome de variável.
(((Dict)))((("tipo", "Base", "Dict", see="Dict")))

[source,@julia-repl-test chap11]
----

> julia> por2sp = Dict()

Reviewer: Será que não faz mais sentido traduzir para "por_para_sp" ou "port_para_esp"?

Dict{Any,Any} with 0 entries
----

> O tipo de dicionário é cercado por chaves: as chaves são do tipo +Any+ e os valores também são do tipo +Any+.

Reviewer: ```suggestion
O tipo do dicionário é cercado por chaves: as chaves são do tipo +Any+ e os valores também são do tipo +Any+.
```

(((chaves)))((("{}", see="chaves")))(((Any)))((("tipo", "Base", "Any", see="Any")))

O dicionário está vazio. Para adicionar items no dicionário, podemos usar os colchetes:
(((colchetes)))

[source,@julia-repl-test chap11]
----
julia> por2sp["um"] = "uno";

----

> Essa linha cria um item que mapeia a chave "um" para o valor "uno". Se nós imprimirmos o dicionário novamente, nós vemos o par chave-valor com a flecha +pass:[=&gt;]+ entre a chave e o valor:

Reviewer: ```suggestion
Essa linha cria um item que mapeia a chave "um" para o valor "uno". Se imprimirmos o dicionário novamente, nós vemos o par chave-valor com a flecha +pass:[=&gt;]+ entre a chave e o valor:
```

(((pass:[=&gt;])))

[source,@julia-repl-test chap11]
----
julia> por2sp
Dict{Any,Any} with 1 entry:
  "um" => "uno"
----

> Esse formato de resultado também é um formato de entrada. Por exemplo, você pode criar um novo dicionário com três itens:

Reviewer: ```suggestion
Essa formatação de saída também é uma formatação de entrada. Por exemplo, você pode criar um novo dicionário com três itens:
```

Reviewer: A tradução de format pode ser formato ou formatação. Mas na minha opinião, eu acho que o termo "formatação" está mais associado ao jeito de escrever do que o termo "formato".

[source,@julia-repl-test chap11]
----
julia> por2sp = Dict("um" => "uno", "dois" => "dos", "três" => "tres")
Dict{String,String} with 3 entries:

>   "dois"   => "dos"

Reviewer: Arrumar o espaçamento, como nas linhas de baixo.

  "um"   => "uno"
  "três" => "tres"
----

Todas as chaves e os valores iniciais são strings, então um +Dict{String, String}+ é criado.

[WARNING]
====

> A ordem do par chave-valor pode não ser a mesma. Se você digitar o mesmo exemplo no seu computador, você pode obter um resultado diferente. Em geral, a ordem dos itens em um dicionário é imprevisível.

Reviewer: Sugestão para tirar repetição de "você":
```suggestion
A ordem do par chave-valor pode não ser a mesma. Se você digitar o mesmo exemplo no seu computador, pode obter um resultado diferente. Em geral, a ordem dos itens em um dicionário é imprevisível.
```

====

> Mas isso não é um problema pois os elementos de um dicionário nunca são indexados com indíces inteiros. Ao invés disso, nós usamos as chaves para consultar os valores correspondentes:

Reviewer: ```suggestion
Mas isso não é um problema pois os elementos de um dicionário nunca são indexados com índices inteiros. Ao invés disso, nós usamos as chaves para consultar os valores correspondentes:
```

[source,@julia-repl-test chap11]
----
julia> por2sp["dois"]
"dos"
----

> A chave "dois" sempre é mapeada par ao valor "dos", então a ordem dos itens não importa.

Reviewer: ```suggestion
A chave "dois" sempre é mapeada para o valor "dos", então a ordem dos itens não importa.
```

Se a chave não está no dicionário, nós recebemos uma exceção:
(((KeyError)))((("error", "Base", "KeyError", see="KeyError")))

[source,@julia-repl-test chap11]
----
julia> por2sp["quatro"]

> ERROR: KeyError: key "four" not found

Reviewer: ```suggestion
ERROR: KeyError: key "quatro" not found
```

----

> A função +length+ funciona em dicionários; ela retorna o número de pares chave-valor:

Reviewer: ```suggestion
A função +length+ funciona nos dicionários e retorna o número de pares chave-valor:
```

Reviewer: Não tem nenhum erro, mas
(1) em "A função +length+ funciona em dicionários", essa mania de comer os artigos (_o, a, os, as_) se chama anglicismo. Tem gente q tem estramento com a leitura qdo isso acontece.
(2) Eu fiz a sugestão que, na minha opinião, melhoram a fluidez da leitura.

(((length)))

[source,@julia-repl-test chap11]
----
julia> length(por2sp)
3
----

A função +keys+ retorna uma lista com as chaves do dicionário:
(((keys)))((("função", "Base", "keys", see="keys")))

[source,@julia-repl-test chap11]
----
julia> ks = keys(por2sp);

julia> print(ks)
["dois", "um", "três"]
----

Agora você pode usar o operador +∈+ para verificar se algo aparece como uma _chave_ no dicionário:

> (((in)))

Reviewer: Eu acho q poderia traduzir para (((operador in))). O q vc acha @joaoluisro ?

Author: Acho que não, dependendo de como foi feito nos outros capítulos é melhor deixar por enquanto.

Reviewer: Blz

[source,@julia-repl-test chap11]
----
julia> "um" ∈ ks
true
julia> "uno" ∈ ks
false
----

> Para verificar se algo aparece como um valor em um dicionário, você pode usar a função +values+, que retorna uma coleção de valores, e em seguida usar o operador +∈+:

Reviewer: ```suggestion
Para verificar se algo aparece como um valor em um dicionário, você pode usar a função +values+ que retorna uma coleção de valores, e em seguida usar o operador +∈+:
```

Reviewer: Sugestão só para melhorar a fluidez da leitura.

(((valores)))((("função", "Base", "valores", see="valores")))

[source,@julia-repl-test chap11]
----
julia> vs = values(por2sp);

julia> "uno" ∈ vs
true
----

> O operador +∈+ utiliza algoritmos diferentes para arrays e dicionários. Para arrays, ele busca os elementos do array em ordem, como em <<searching>>. À medida que o array fica maior, o tempo de busca cresce em proporção direta.

Reviewer: ```suggestion
O operador +∈+ utiliza algoritmos diferentes para os arrays e dicionários. Para os arrays, ele busca os elementos do array em ordem, como em <<searching>>. À medida que o array fica maior, o tempo de busca cresce diretamente proporcional.
```

Reviewer: São só sugestões... não tem nenhum erro.

> Para dicionários, o Julia usa um algoritmo chamado de _tabela hash_ que possui uma propriedade expecional: o operador +∈+ leva aproximadamente o mesmo não importa quantos itens o dicionário possui.

Reviewer: ```suggestion
Para os dicionários, o Julia usa um algoritmo chamado _tabela hash_ (ou tabela de dispersão ou tabela de espalhamento) que possui uma propriedade excepcional: o operador +∈+ leva aproximadamente o mesmo tempo independente da quantidade de itens do dicionário.
```

(((tabela hash)))

[[dictionary_collection_counters]]
=== Dicionários como uma Coleção de Contadores

> Suponha que é dado à você uma string e você gostaria de contar quantas vezes cada letra aparece. Existem muitas maneiras de fazer isso:

Reviewer: Sugestão 1: 
```suggestion
Suponha que é dado a você uma string e você gostaria de contar quantas vezes cada letra aparece. Existem muitas maneiras de fazer isso:
```

> * Você poderia criar 26 variáveis, um para cada letra do alfabeto. Em seguida você poderia percorrer a string, e para cada caractere, incrementar o contador correspondente, provávelmente usando um condicional encadeado.

Reviewer: ```suggestion
* Você poderia criar 26 variáveis, um para cada letra do alfabeto. Em seguida você poderia percorrer a string, e para cada caractere, incrementar o contador correspondente, provavelmente usando um condicional encadeado.
```

> * Você poderia criar um array com 26 elementos. Em seguida você poderia converter cada caractere para um número (usando a função embutida +Int+), usar o número como um indíce para o array, e incrementar o contador adequado.

Reviewer: Sugestão para tirar repetição do você na 1a frase e logo em seguida na 2a frase.
```suggestion
* Você poderia criar um array com 26 elementos. Em seguida pode-se converter cada caractere para um número (usando a função embutida +Int+), usar o número como um índice para o array, e incrementar o contador adequado.
```

> * Você poderia criar um dicionário com caracteres como chaves e contadores como seus valores correspondentes. A primeira vez que você vê um caractere, você adicionaria um item no dicionário. Após isso você incrementaria o valor de um item existente.

Reviewer: Minha sugestão para melhorar a leitura do texto.
```suggestion
* Você poderia criar um dicionário com os caracteres como as chaves e os contadores como seus valores correspondentes. A primeira vez que você vê um caractere, adicionaria um item no dicionário. Após isso, você só incrementaria o valor de um item existente.
```

> Cada uma dessas opções efetua o mesmo cálculo, mas cada uma delas implementa este cálculo de maneiras diferentes.

Reviewer: Vc tem usado mto "maneira". Como vc já vai usar o termo "maneira" no próximo parágrafo, vamos dar uma variada.
```suggestion
Cada uma dessas opções efetua o mesmo cálculo, mas cada uma delas implementa este cálculo de formas diferentes.
```

> Uma _implementação_ é uma maneira de efetuar um cálculo; algumas implementações são melhores que outras. Por exemplo, uma vantagem da implementação do dicionário é que nós não temos que saber antecipadamente quais letras aparecem na string, e nós temos que somente criar espaço para as letras que aparecem.

Reviewer: Minha sugestão para melhorar a fluidez da leitura.
```suggestion
Uma _implementação_ é uma maneira de efetuar um cálculo e algumas implementações são melhores que outras. Por exemplo, uma vantagem da implementação do dicionário é que nós não temos que saber antecipadamente quais letras aparecem na string, e sim criar espaço só para as letras que aparecem.
```

[[chap11]]
== Dicionários

> Este capítulo apresenta outra estrutura de dados embutida chamado dicionário.

Reviewer 2: ```suggestion
Este capítulo apresenta outra estrutura de dados embutida chamada dicionário.
```

=== Um Dicionário é um Mapeamento

> Um _dicionário_ é como uma lista, mas mais geral. Em uma lista, os índices devem ser inteiros e em um dicionário, eles podem (quase) ser de qualquer tipo.

Reviewer 2: ```suggestion
Um _dicionário_ é como uma lista, mas mais geral. Em uma lista, os índices devem ser inteiros e em um dicionário, eles podem ser de (quase) qualquer tipo.
```

(((dicionário)))

Um dicionário contém uma coleção de índices, que são chamados de _chaves_, e uma coleção de valores. Cada chave é associada a um único valor. A associação de uma chave e um valor é chamada de _par chave-valor_ ou, às vezes, de item.
(((chave)))(((valor)))(((par chave-valor)))(((item)))

> Na linguagem matemática, um dicionário representa um _mapeamento_ da chave para os valores, então você também pode dizer que cada chave “é mapeado” para um valor. Como um exemplo, iremos construir um dicionário que mapeia palavras em Português para palavras em Espanhol, com as chaves e os valores todos strings.

Reviewer: ```suggestion
Na linguagem matemática, um dicionário representa um _mapeamento_ das chaves para os valores, então você também pode dizer que cada chave “é mapeado” para um valor. Como um exemplo, iremos construir um dicionário que mapeia palavras em Português para palavras em Espanhol, com as chaves e os valores todos strings.
```

> Na linguagem matemática, um dicionário representa um _mapeamento_ das chaves para os valores, então você também pode dizer que cada chave “é mapeado” para um valor. Como um exemplo, iremos construir um dicionário que mapeia palavras em Português para palavras em Espanhol, com as chaves e os valores todos strings.

Reviewer 2: ```suggestion
Na linguagem matemática, um dicionário representa um _mapeamento_ das chaves para os valores, então você também pode dizer que cada chave “é mapeada” para um valor. Como um exemplo, iremos construir um dicionário que mapeia palavras em Português para palavras em Espanhol, com as chaves e os valores todos strings.
```

(((mapeamento)))

A função +Dict+ cria um novo dicionário sem items. Como +Dict+ é um nome de uma função embutida, devemos evitar usar ele como um nome de variável.
(((Dict)))((("tipo", "Base", "Dict", see="Dict")))

[source,@julia-repl-test chap11]
----
julia> por_para_esp = Dict()
Dict{Any,Any} with 0 entries
----

O tipo do dicionário é cercado por chaves: as chaves são do tipo +Any+ e os valores também são do tipo +Any+.
(((chaves)))((("{}", see="chaves")))(((Any)))((("tipo", "Base", "Any", see="Any")))

O dicionário está vazio. Para adicionar items no dicionário, podemos usar os colchetes:
(((colchetes)))

[source,@julia-repl-test chap11]
----

> julia> por_para_esp["um"] = "uno";

Reviewer 2: ```suggestion
julia> pt_para_esp["um"] = "uno";
```

----

Essa linha cria um item que mapeia a chave "um" para o valor "uno". Se imprimirmos o dicionário novamente, nós vemos o par chave-valor com a flecha +pass:[=&gt;]+ entre a chave e o valor:
(((pass:[=&gt;])))

[source,@julia-repl-test chap11]
----

> julia> por_para_esp

Reviewer 2: ```suggestion
julia> pt_para_esp
```

Dict{Any,Any} with 1 entry:
  "um" => "uno"
----

Essa formatação de saída também é uma formatação de entrada. Por exemplo, você pode criar um novo dicionário com três itens:

[source,@julia-repl-test chap11]
----

> julia> por_para_esp = Dict("um" => "uno", "dois" => "dos", "três" => "tres")

Reviewer 2: ```suggestion
julia> pt_para_esp = Dict("um" => "uno", "dois" => "dos", "três" => "tres")
```

Dict{String,String} with 3 entries:
  "dois" => "dos"
  "um"   => "uno"
  "três" => "tres"
----

> Todas as chaves e os valores iniciais são strings, então um +Dict{String, String}+ é criado.

Reviewer: ```suggestion
Todas as chaves e os valores iniciais são strings, então um +Dict{String,String}+ é criado.
```

[WARNING]
====
A ordem do par chave-valor pode não ser a mesma. Se você digitar o mesmo exemplo no seu computador, pode obter um resultado diferente. Em geral, a ordem dos itens em um dicionário é imprevisível.
====

Mas isso não é um problema pois os elementos de um dicionário nunca são indexados com índices inteiros. Ao invés disso, nós usamos as chaves para consultar os valores correspondentes:

[source,@julia-repl-test chap11]
----

> julia> por_para_esp["dois"]

Reviewer 2: ```suggestion
julia> pt_para_esp["dois"]
```

"dos"
----

A chave "dois" sempre é mapeada para o valor "dos", então a ordem dos itens não importa.

Se a chave não está no dicionário, nós recebemos uma exceção:
(((KeyError)))((("error", "Base", "KeyError", see="KeyError")))

[source,@julia-repl-test chap11]
----

> julia> por_para_esp["quatro"]

Reviewer 2: ```suggestion
julia> pt_para_esp["quatro"]
```

ERROR: KeyError: key "quatro" not found
----

A função +length+ funciona nos dicionários e retorna o número de pares chave-valor:
(((length)))

[source,@julia-repl-test chap11]
----

> julia> length(por_para_esp)

Reviewer 2: ```suggestion
julia> length(pt_para_esp)
```

3
----

A função +keys+ retorna uma lista com as chaves do dicionário:
(((keys)))((("função", "Base", "keys", see="keys")))

[source,@julia-repl-test chap11]
----

> julia> ks = keys(por_para_esp);

Reviewer 2: ```suggestion
julia> ks = keys(pt_para_esp);
```

julia> print(ks)
["dois", "um", "três"]
----

Agora você pode usar o operador +∈+ para verificar se algo aparece como uma _chave_ no dicionário:
(((in)))

[source,@julia-repl-test chap11]
----
julia> "um" ∈ ks
true
julia> "uno" ∈ ks
false
----

Para verificar se algo aparece como um valor em um dicionário, você pode usar a função +values+ que retorna uma coleção de valores, e em seguida usar o operador +∈+:

> (((valores)))((("função", "Base", "valores", see="valores")))

Reviewer: Eu acho q neste caso não precisa traduzir para valores porque estamos falando da função embutida values.
```suggestion
(((values)))((("função", "Base", "values", see="values")))
```

[source,@julia-repl-test chap11]
----

> julia> vs = values(por_para_esp);

Reviewer 2: ```suggestion
julia> vs = values(pt_para_esp);
```

end
----

O nome da função é +histograma+, que é um termo estátistico para uma coleção de contadores (ou frequências).
(((histograma)))((("função", "definido pelo programador", "histograma", see="histograma")))

> A primeira linha da função cria um dicionário vazio. O loop +for+ percorre a string. Toda vez que o loop é percorrido, se o caractere +c+ não está no dicionário, nós criamos um novo item com a chave +c+ e o valor inicial +1+ (já que nós vimos esta letra uma vez). Se +c+ já está no dicionário, nós incrementamos +d[c]+.

Reviewer: Eu tinha a impressão de que combinamos de usar laço como tradução de loop.

Funciona da seguinte forma:

[source,@julia-repl-test chap11]
----

> julia> h = histograma("brontosaurus")

Reviewer: ```suggestion
julia> h = histograma("brontossauro")
```

  'b' => 1

> ----

Reviewer: Arrumar a saída do dicionário tb após mudar para "brontossauro".

> O histograma indica que as letras +'a'+ e +'b'+ aparecem uma vez; +'o'+ aparece duas, e assim em diante.

Reviewer: ```suggestion
O histograma indica que as letras +'a'+ e +'b'+ aparecem uma vez; +'o'+ aparece três, e assim em diante.
```

> Dicionários possuem uma função chamada +get+ que recebe uma chave e uma valor padrão. Se a chave aparece no dicionário, +get+ retorna o valor correspondente; caso contrário ela retorna o valor padrão. Por exemplo:

Reviewer: ```suggestion
Dicionários possuem uma função chamada +get+ que recebe uma chave e um valor padrão. Se a chave aparece no dicionário, +get+ retorna o valor correspondente; caso contrário ela retorna o valor padrão. Por exemplo:
```

0
----

===== Exercício 11-1

Use +get+ para escrever +histograma+ de uma maneira mais concisa. Você deve ser capaz de eliminar a declaração +if+.

> === Dando Laços e Dicionários

Reviewer: ```suggestion
=== Laços e Dicionários
```

Você pode percorrer as chaves de um dicionário em uma declaração +for+. Por exemplo, +imprimehist+ exibe cada chave e o seu valor correspondente:
(((imprimehist)))((("função", "definido pelo programador", "imprimehist", see="imprimehist")))

[source,@julia-setup chap11]
----
function imprimehist(h)
    for c in keys(h)
        println(c, " ", h[c])
    end
end
----

Aqui está o resultado:

[source,@julia-repl-test chap11]
----
julia> h = histograma("papagaio");

julia> imprimehist(h)
a 3
g 1
p 2
i 1
o 1
----

> Novamente, as chaves não estão em nenhuma ordem específica. Para percorrer as chaves em ordem, você pode combinar +sort+ e +collect+:

Reviewer: ```suggestion
Novamente as chaves não estão em nenhuma ordem específica. Para percorrer as chaves em ordem, você pode combinar +sort+ e +collect+:
```

(((sort)))(((collect)))

[source,@julia-repl-test chap11]
----
julia> for c in sort(collect(keys(h)))
           println(c, " ", h[c])
       end
a 3
g 1
i 1
o 1
p 2
----

> === Consulta Invertida

Reviewer: Minha sugestão:
```suggestion
=== Consulta Inversa
```

Reviewer: Se vc aderir à sugestão, acho legal vc trocar todos os termos "consulta invertida" para "consulta inversa" e não ficar só no título.

Dado um dicionário +d+ e uma chave +k+, é fácil achar o valor correspondente +v = d[k]+. Esta operação é chamada de _consulta_.
(((consulta)))

> Mas e se você tem +v+ e quer achar +k+? Você tem dois problemas: primeiro, pode haver mais de uma chave que mapeia para o valor +v+. Dependendo da aplicação, você poderia escolher um, ou você teria que criar um array que contém todos eles. Segundo, não há uma sintaxe simples que faz uma _consulta invertida_ ; você tem que procurar.

Reviewer: ```suggestion
Mas e se você tem +v+ e quer achar +k+? Você tem dois problemas: primeiro, pode haver mais de uma chave que mapeia para o valor +v+. Dependendo do que é pedido, você poderia escolher um, ou teria que criar um array que contém todos eles. Segundo, não há uma sintaxe simples que faz uma _consulta invertida_; você tem que procurar.
```

(((consulta invertida)))

Aqui está uma função que recebe um valor e retorna a primeira chave que mapeia a este valor:

[source,@julia-setup chap11]
----
function consultainvertida(d, v)
    for k in keys(d)
        if d[k] == v
            return k
        end
    end
    error("Erro de Consulta")
end
----

Esta função é mais um exemplo do padrão de busca, mas usa uma função que ainda não vimos, +error+. A função +error+ é usada para gerar um +ErrorException+ que interrompe o fluxo normal de controle. Neste caso ela tem a mensagem +"Erro de Consulta"+, indicando que a chave não existe.
(((error)))((("função", "Base", "error", see="error")))

Se nós chegarmos no final do loop, isso significa que +v+ não aparece no dicionário como um valor, então geramos uma exceção.

Aqui está um exemplo de uma consulta invertida bem-sucedida:

[source,@julia-repl-test chap11]
----
julia> h = histograma("papagaio");

julia> chave = consultainvertida(h, 2)
'p': ASCII/Unicode U+0072 (category Ll: Letter, lowercase)
----

E uma malsucedida:

[source,@julia-repl-test chap11]
----
julia> chave = reverselookup(h, 4)

> ERROR: LookupError

Reviewer: Faltou traduzir o "reverselookup" e "LookupError".

----

> O efeito de uma exceção gerada é o mesmo quando o Julia gera um: ele exibe o stracktrace e uma mensagem de erro.

Reviewer: ```suggestion
O resultado de uma exceção gerada é o mesmo quando o Julia gera um: ele exibe o stracktrace e uma mensagem de erro.
```

Julia fornece uma maneira otimizada de fazer uma consulta invertida: +findall(isequal(3), h)+.
(((findall)))((("função", "Base", "findall", see="findall")))(((isequal)))((("função", "Base", "isequal", see="isequal")))

[WARNING]
====

> Uma consulta invertida é muito  mais devagar que uma consulta normal; se você tiver que executá-la várias vezes, ou se o dicionário ficar muito grande, a performance do seu programa irá diminiuir.

Reviewer: ```suggestion
Uma consulta invertida é muito mais demorada que uma consulta normal; se você tiver que executá-la várias vezes, ou se o dicionário ficar muito grande, o desempenho do seu programa diminuirá.
```

====

=== Dicionários e Arrays

> Arrays podem aparecer como valores em um dicionário. Por exemplo, se você receber um dicionário que mapeia letras à frequências, você pode querer invertê-lo; isto é, criar um dicionário que mapeia frequências para letras. Já que várias letras podem ter a mesma frequência, cada valor em um dicionário invertido deve ser um array de letras.

Reviewer: ```suggestion
Arrays podem aparecer como valores em um dicionário. Por exemplo, se você receber um dicionário que mapeia letras às frequências, você pode querer invertê-lo; isto é, criar um dicionário que mapeia frequências até as letras. Já que várias letras podem ter a mesma frequência, cada valor em um dicionário invertido deve ser um array de letras.
```

0
----

===== Exercício 11-1

Use +get+ para escrever +histograma+ de uma maneira mais concisa. Você deve ser capaz de eliminar a declaração +if+.

=== Laços e Dicionários

> Você pode percorrer as chaves de um dicionário em uma declaração +for+. Por exemplo, +imprimehist+ exibe cada chave e o seu valor correspondente:

Reviewer: Vc vai deixar "imprime_hist" ou "imprimehist" mesmo?

(((consulta inversa)))

Aqui está uma função que recebe um valor e retorna a primeira chave que mapeia a este valor:

[source,@julia-setup chap11]
----
function consulta_inversa(d, v)
    for k in keys(d)
        if d[k] == v
            return k
        end
    end
    error("Erro de Consulta")
end
----

Esta função é mais um exemplo do padrão de busca, mas usa uma função que ainda não vimos, +error+. A função +error+ é usada para gerar um +ErrorException+ que interrompe o fluxo normal de controle. Neste caso ela tem a mensagem +"Erro de Consulta"+, indicando que a chave não existe.
(((error)))((("função", "Base", "error", see="error")))

Se nós chegarmos no final do laço, isso significa que +v+ não aparece no dicionário como um valor, então geramos uma exceção.

Aqui está um exemplo de uma consulta inversa bem-sucedida:

[source,@julia-repl-test chap11]
----
julia> h = histograma("papagaio");

julia> chave = consulta_inversa(h, 2)
'p': ASCII/Unicode U+0072 (category Ll: Letter, lowercase)
----

E uma malsucedida:

[source,@julia-repl-test chap11]
----
julia> chave = consulta_inversa(h, 4)
ERROR: Erro de Consulta
----

> O resultado de uma exceção gerada é o mesmo quando o Julia gera um: ele exibe o stracktrace e uma mensagem de erro.

Reviewer 2: ```suggestion
O resultado de uma exceção gerada é o mesmo quando o Julia gera um: ele exibe o stacktrace e uma mensagem de erro.
```

            inverso[valor] = [chave]
        else
            push!(inverso[valor], chave)
        end
    end
    inverso
end
----

Cada vez que o laço é percorrido, +chave+ recebe uma chave de +d+ e +valor+ recebe o valor correspondente. Se +valor+ não está em +inverso+, isto significa que não a vimos ainda, então criamos um novo item e inicializamos com um _singleton_ (uma lista que contém um único elemento). Caso contrário esse valor já foi visto, e então acrescentamos a chave correspondente à lista.
(((singleton)))

Aqui está um exemplo:

[source,@julia-repl-test chap11]
----
julia> hist = histograma("papagaio");

julia> inverso = inverte_dict(hist)
Dict{Any,Any} with 2 entries:
  3 => ['a']
  2 => ['p']
  1 => ['g', 'i', 'o']
----

[[fig11-1]]
.Diagrama de Estado
image::images/fig111.svg[]

<<fig11-1>> é um diagrama de estado mostrando +hist+ e +inverso+. Um dicionário é representado como uma caixa com os pares chave-valor dentro. Para os valores que são inteiros, pontos flutuantes ou strings, eu os desenho dentro da caixa, já para as listas normalmente desenho fora da caixa, só para simplificar o diagrama.
(((diagrama de estado)))

[NOTE]
====

> Eu mencionei anteriormente que um dicionário é implementado usando uma tabela hash e isso significa que as chaves devem ser _hashable_, ou de um tipo que permite que uma função hash atuem sobre elas.

Reviewer 2: ```suggestion
Mencionamos anteriormente que um dicionário é implementado usando uma tabela hash e isso significa que as chaves devem ser _hashable_, isto é, de um tipo que permite que uma função hash atuem sobre elas.
```

(((hashable)))

Uma _hash_ é uma função que recebe um valor (de qualquer tipo) e retorna um inteiro. Dicionários usam estes inteiros, chamados de valores hash, para guardar e consultar pares chave-valor.
(((hash)))
====

[[memos]]
=== Memos

> Se você já brincou com a função +fibonacci+ de <<one_more_exemple>>, você pode ter percebido que quanto maior o argumento que você fornece, mais tempo a função leva para executar. Além disso, o tempo de execução cresce rapidamente.

Reviewer: Sugestão para tirar um pouco da repetição de "você".
```suggestion
Se você já brincou com a função +fibonacci+ de <<one_more_exemple>>, pode ter percebido que quanto maior o argumento que você fornece, mais tempo a função leva para executar. Além disso, o tempo de execução cresce rapidamente.
```

(((fibonacci)))

Para entender o porquê, considere <<fig11-2>>, que mostra um _grafo de chamada_ para +fibonacci+ com +n = 4+:
(((grafo de chamada)))((("diagram", "grafo de chamada", see="grafo de chamada")))

[[fig11-2]]
.Grafo de chamada
image::images/fig112.svg[]

Um grafo de chamada mostra um conjunto de quadros da função, com linhas conectando cada quadro aos quadros que a função chama. No topo do grafo, +fibonacci+ com +n = 4+ chama +fibonacci+ com +n = 3+ e +n = 2+. Por sua vez, +fibonacci+ com +n = 3+ chama +fibonacci+ com +n = 2+ e +n = 1+. E assim em diante.

Conte quantas vezes +fibonacci(0)+ e +fibonacci(1)+ são chamadas. Está é uma solução ineficiente do problema, e fica pior à medida que o argumento aumenta.

Uma solução é acompanhar os valores já calculados armazenando-os em um dicionário. Um valor previamente calculado que é armazenado para uso posterior é chamado de _memo_. Aqui está uma versão “memoizada” de fibonacci:
(((memo)))

[source,@julia-setup]
----
conhecidos = Dict(0=>0, 1=>1)

function fibonacci(n)
    if n ∈ keys(conhecidos)
        return conhecidos[n]
    end
    res = fibonacci(n-1) + fibonacci(n-2)
    conhecidos[n] = res
    res
end
----

> +conhecido+ é um dicionáro que guarda os números de Fibonacci que já sabemos. Ele começa com dois itens: +0+ mapeia para +0+ e +1+ mapeia para +1+.

Reviewer 2: ```suggestion
+conhecidos+ é um dicionáro que guarda os números de Fibonacci que já sabemos. Ele começa com dois itens: +0+ mapeia para +0+ e +1+ mapeia para +1+.
```

Toda vez que +fibonacci+ é chamada, ela checa +conhecidos+. Se o resultado já está lá, ela retorna imediatamente. Por outro lado ela tem que computar um novo valor, adicionar ele ao dicionário, e retorná-lo.

Se você executar esta versão de +fibonacci+ e comparar com a original, você verá que a atual é muito mais rápida.

[[global_variables]]
=== Variáveis Globais

No exemplo anterior, conhecidos é criado fora da função, então ela pertence ao quadro especial chamado +Main+. Variáveis em +Main+ são às vezes chamadas de _globais_ pois podem ser acessadas de qualquer função. Diferente de variáveis locais, que desaparecem quando a função acaba, variáveis globais persistem de uma chamada de função para a próxima.
(((global)))((("palavra-chave", "global", see="global")))(((variável global)))

> É comum usar variáveis globais para _flags_; isto é, variáveis booleanas que indicam (“sinal”) se uma condição é verdadeira. Por exemplo, alguns programas usam uma flag chamada +verbose+ para controlar o nível de detalhamento na saída:

Reviewer 2: ```suggestion
É comum usar variáveis globais para _flags_; isto é, variáveis booleanas que indicam (“sinalizam”) se uma condição é verdadeira. Por exemplo, alguns programas usam uma flag chamada +verbose+ para controlar o nível de detalhamento na saída:
```

(((flag)))

[source,@julia-setup]
----
verbose = true

function exemplo1()
    if verbose
        println("Executando exemplo1")
    end
end
----

> Se tentar reatribuir uma variável global, você poderá se surpreender. O exemplo a seguir deve acompanhar se a função foi chamada:

Reviewer: ```suggestion
Se você tentar reatribuir uma variável global, poderá se surpreender. O exemplo a seguir deve acompanhar se a função foi chamada:
```

Aqui está um exemplo que tenta atualizar uma variável global:

[source,@julia-setup chap11]
----
conta = 0

function exemplo3()
    conta = conta + 1          # ERRADO
end
----

Ao executar a função, você recebe:
(((UndefVarError)))

[source,@julia-repl-test chap11]
----
julia> exemplo3()
ERROR: UndefVarError: conta not defined
----

O Julia assume que +conta+ é local, partindo da suposição de que você está lendo a função antes de escrevê-la. A solução, novamente, é declarar +conta+ como global.

[source,@julia-setup]
----
conta = 0

function exemplo3()
    global conta
    conta += 1
end
----

Se uma variável global se refere a um valor mutável, você pode modificar o valor sem declarar a variável global:

[source,@julia-setup]
----
conhecido = Dict(0=>0, 1=>1)

function exemplo4()
    conhecido[2] = 1
end
----

Então você pode adicionar, remover e substituir os elementos de um array global ou dicionário, mas se você quiser reatribuir a variável, você deve declará-la como global:

[source,@julia-setup]
----
conhecido = Dict(0=>0, 1=>1)

function exemplo5()
    global conhecido
    conhecido = Dict()
end
----

> Por razões de performance, deve-se declarar uma variável global como constante. Não é mais possível reatribuir a variável mas caso ela faça referência a um valor mutável, você pode modificar o valor.

Reviewer: ```suggestion
Por razões de performance, deve-se declarar uma variável global como constante. Você já não pode reatribuir a variável mas caso ela faça referência a um valor mutável, pode-se modificar o valor.
```

(((const)))((("palavra-chave", "const", see="const")))(((variável global constante)))

[source,@julia-setup]
----
const conhecido = Dict(0=>0, 1=>1)

function exemplo4()
    conhecido[2] = 1
end
----

[WARNING]
====

> Variáveis globais podem ser utéis, mas se tem muitas delas, e você modifica-as frequentemente, elas podem ser a causa dos programas serem dificéis de debugar e terem mau desempenho.

Reviewer: ```suggestion
Variáveis globais podem ser utéis, mas se tem muitas delas, e você modifica-as frequentemente, elas podem ser a causa dos programas serem dificéis de depurar e terem mau desempenho.
```

====

> === Debugando

Reviewer: ```suggestion
=== Depuração
```

> Na medida em que você trabalha com datasets maiores, pode ser que seja difícil depurar imprimindo e checando a saída na mão. Aqui estão algumas sugestões para depurar datasets maiores:

Reviewer 2: ```suggestion
Na medida em que você trabalha com conjuntos de dados maiores, pode ser que seja difícil depurar imprimindo e checando a saída na mão. Aqui estão algumas sugestões para depurar conjuntos de dados maiores:
```

> Na medida em que você trabalha com datasets maiores, pode ser que seja difícil debugar imprimindo e checando a saída na mão. Aqui estão algumas sugestões para debugar datasets maiores:

Reviewer: ```suggestion
Na medida em que você trabalha com datasets maiores, pode ser que seja difícil depurar imprimindo e checando a saída na mão. Aqui estão algumas sugestões para depurar datasets maiores:
```

> (((debugando)))

Reviewer: ```suggestion
(((depuração)))
```

* Diminua a entrada:
+

> Se possível, reduza o tamanho do dataset. Por exemplo, se o programa lê um arquivo de texto, comece com apenas as 10 primeiras linhas, ou com o menor exemplo que dá erro. Você não deve editar os arquivos em si, mas modificar o programa para que ele leia somente as primeiras latexmath:[\(n\)] linhas.

Reviewer 2: ```suggestion
Se possível, reduza o tamanho do conjunto de dados. Por exemplo, se o programa lê um arquivo de texto, comece com apenas as 10 primeiras linhas, ou com o menor exemplo que dá erro. Você não deve editar os arquivos em si, mas modificar o programa para que ele leia somente as primeiras latexmath:[\(n\)] linhas.
```

(((redução de tamanho)))
+
Se existe um erro, você pode reduzir de latexmath:[\(n\)] para o menor valor que dá erro, e em seguida incrementá-lo gradualmente à medida que você encontra e corrige os erros.

* Verifique os resumos e os tipos:
+

> Ao invés de imprimir e checar o dataset inteiro, considere imprimir os resumos dos dados: por exemplo, o número de itens em um dicionário ou o total de uma lista de números.

Reviewer 2: ```suggestion
Ao invés de imprimir e checar o conjunto de dados inteiro, considere imprimir os resumos dos dados: por exemplo, o número de itens em um dicionário ou o total de uma lista de números.
```

(((verificação)))
+

> Uma causa comum de erros de execução é um valor que não é do tipo correto. Para debugar esse tipo de erro, geralmente é suficiente imprimir o tipo de um valor.

Reviewer: ```suggestion
Uma causa comum de erros de execução é um valor que não é do tipo correto. Para a depuração desse tipo de erro, geralmente é suficiente imprimir o tipo de um valor.
```

* Escreva auto-verificações:
+
Algumas vezes você pode escrever código para checar os erros automaticamente. Por exemplo, se você está calculando a média de um array de números, você poderia checar que o resultado não está acima do maior elemento do array ou abaixo do menor elemento. Isso é chamado de “verificação de sanidade”.
(((auto-verificação)))(((verificação de sanidade)))
+
Outro tipo de verificação compara o resultado de dois cálculos diferentes para verificar se eles são consistentes. Isso é chamado de “verificação de consistência”.
(((verificação de consistência)))

* Formate a saída:
+

> Resultados de debugging formatados podem facilitar a detecção de erros, como visto em um exemplo em <<deb06>>.

Reviewer: ```suggestion
Resultados de depuração formatados podem facilitar a detecção de erros, como visto em um exemplo em <<deb06>>.
```

(((formatação)))
+

> E mais uma vez, o tempo que você usa construindo andaimes pode reduzir o tempo gasto no debugging.

Reviewer: ```suggestion
E mais uma vez, o tempo que você usa construindo andaimes pode reduzir o tempo gasto na depuração.
```

> (((andaime)))(((debugging)))

Reviewer: ```suggestion
(((andaime)))(((depuração)))
```

=== Glossário

mapeamento::
Uma relação na qual cada elemento de um conjunto corresponde a um elemento de outro conjunto.
(((mapeamento)))

dicionário::
Um mapeamento de chaves para os seus valores correspondentes.
(((dicionário)))

par chave-valor::
A representação de um mapeamento de uma chave para um valor.
(((par chave-valor)))

item::
Em um dicionário, é outro nome para o par chave-valor.
(((item)))

chave::
Um objeto que aparece em um dicionário como a primeira parte de um par chave-valor.
(((chave)))

valor::
Um objeto que aparece em um dicionário como a segunda parte de um par chave-valor. Isso é mais específico que o nosso uso prévio da palavra “valor”.
(((valor)))

implementação::

> Uma maneira de efetuar cálculos.

Reviewer: ```suggestion
Uma maneira de efetuar os cálculos.
```

(((implementação)))

tabela hash::

> O algoritmo usado para implementar dicionários em Julia.

Reviewer: ```suggestion
O algoritmo usado para implementar os dicionários em Julia.
```

(((consulta)))

consulta inversa::
Uma operação em um dicionário que recebe um valor e encontra uma ou mais chaves mapeadas para ele.
(((consulta inversa)))

singleton::
Um array (ou outra sequência) com um único elemento.
(((singleton)))

grafo de chamada::
Um diagrama que mostra todo quadro criado durante a execução de um programa, com uma flecha que vai de quem chama para quem é chamado.
(((grafo de chamada)))

memo::
Um valor já computado e guardado para evitar cálculos futuros desnecessários.
(((memo)))

variável global::
Uma variável definida fora da função. Variáveis globais podem ser acessadas de qualquer função.
(((variável global)))

declaração global::
Uma declaração para tornar um nome de variável como global.
(((declaração global)))

flag::
Uma variável booleana usada para indicar se uma condição é verdadeira.
(((flag)))

declaração::
Uma declaração como +global+ que informa ao interpretador algo sobre a variável.
(((declaração)))

variável global constante::
Uma variável global que não pode ser reatribuída.
(((variável global constante)))

=== Exercícios

[[ex11-1]]
===== Exercício 11-2

> Escreva uma função que leia palavras em _palavras.txt_ e guarde-as como chaves em um dicionário. Não importa quais sejam os valores. Em seguida, você pode usar o operador +∈+ como uma maneira rápida de verificar se uma string está ou não no dicionário.

Reviewer: ```suggestion
Escreva uma função que leia as palavras em _palavras.txt_ e guarde-as como chaves em um dicionário. Não importa quais sejam os valores. Em seguida, você pode usar o operador +∈+ como uma maneira rápida de verificar se uma string está ou não no dicionário.
```

(((in)))

Se o exercício <<ex10-10>> foi feito, você pode comparar a velocidade desta implementação com o operador +∈+ em arrays e na busca em bissecção.

[[ex11-2]]
===== Exercício 11-3

Leia a documentação da função de dicionário +get!+ e use-a para escrever uma versão mais concisa de +inverte_dict+.
(((get!)))((("função", "Base", "get!", see="get!")))

[[ex11-3]]
===== Exercício 11-4

Memoize a função de Ackermann de <<ex06-2>> e verifique se a memoização possibilita a avaliação da função com argumentos maiores.
(((memo)))(((função de Ackermann)))

[[ex11-4]]
===== Exercício 11-5

Se <<ex10-7>> foi feito, então já possui uma função chamada +tem_duplas+ que recebe um array como parâmetro e retorna +true+ se há qualquer objeto que aparece mais de uma vez no array.
(((tem_duplas)))

Use um dicionário para escrever uma versão mais rápida e simplificada de +tem_duplas+.

[[ex11-5]]
===== Exercício 11-6

> Duas palavras são “pares rotacionados” se você pode rotacionar um deles e obter o outro (ver +rotacionapalavra+ em <<ex08-5>>).

Reviewer: Vc vai manter "rotaciona_palavra" ou "rotacionapalavra"? Comentário válido para linha 679.

(((rotacionapalavra)))

Escreva um programa que lê um array e encontra todos os pares rotacionados.

[[ex11-6]]
===== Exercício 11-7

Aqui está outro quebra cabeça de Car Talk (https://www.cartalk.com/puzzler/browse):
(((Car Talk)))

[quote]
____

> Essa foi enviada por um sujeito chamado Dan O’Leary. ele encontrou uma palavra comum de uma sílaba e cinco letras recentemente que possui a seguinte propriedade. Quando você remove a primeira letra, as letras restantes formam um homófono da palavra original, isto é, uma palavra que soa exatamente igual. Substituia a primeira letra, isto é, coloque-a novamente e remova a segunda letra, e o resultado é outro homófono da palavra original. E a pergunta é, qual é a palavra?

Reviewer: ```suggestion
Essa foi enviada por um sujeito chamado Dan O’Leary. Ele encontrou recentemente uma palavra em inglês comum de uma sílaba e cinco letras, que possui a seguinte propriedade peculiar. Quando você remove a primeira letra, as letras restantes formam um homófono da palavra original, isto é, uma palavra que soa exatamente igual. Troque a primeira letra, isto é, coloque-a novamente e remova a segunda letra, e o resultado é outro homófono da palavra original. E a pergunta é, qual é a palavra?
```

> Agora irei dar um exemplo que não funciona. Vamos olhar para uma palavra de cinco letras, ‘wrack.’ W-R-A-C-K, sabe como ‘wrack with pain.’ Se você remover a primeira letra, tenho uma palavra de quatro letras, ’R-A-C-K.’ como em, ‘Holy cow, did you see the rack on that buck! It must have been a nine-pointer!’ é um homófono perfeito. Se você colocar o ‘w’ novamente, e remover o ‘r,’ ao invés disso, você tem uma palavra, ‘wack,’ que é uma palavra real, só não é um homófono das outras duas palavras.

Reviewer: ```suggestion
Agora irei dar um exemplo que não funciona. Vamos olhar para uma palavra de cinco letras, ‘wrack.’ W-R-A-C-K, como na expressão ‘wrack with pain.’ Se eu remover a primeira letra, tenho uma palavra de quatro letras, ’R-A-C-K.’ Como em, ‘Holy cow, did you see the rack on that buck! It must have been a nine-pointer!’ É um homófono perfeito. Se você colocar o ‘w’ novamente, e remover o ‘r’, você fica com a palavra ‘wack,’ que é uma palavra real, só não é um homófona das outras duas palavras.
```

> Mas há, no entanto, pelo menos uma palavra que Dan e nós conhecemos, que irá produzir dois homófonos se você remover tanto as primeiras duas letras para criar duas novas palavras de quatro letras. A pergunta é, qual é a palavra?

Reviewer: Eu sugeri tirar o termo "no entanto" porque já existe o mas. Como mas e no entanto são conjuções que passa a mesma ideia, de contradição, acabei retirando um. Na minha opinião, se aproxima mais do português.
```suggestion
Mas há pelo menos uma palavra, que Dan e nós conhecemos, que irá produzir dois homófonos se você remover tanto as duas primeiras letras para criar duas novas palavras de quatro letras. A pergunta é, qual é a palavra?
```

____

Você pode usar o dicionário de <<ex11-1>> para verificar se uma string está no array de palavras.

[TIP]
====

> Para verificar se duas palavras em inglês são homônimas, você pode usar o Dicionário CMU de Pronunciação. E também pode baixá-lo em  http://www.speech.cs.cmu.edu/cgi-bin/cmudict.

Reviewer: Desculpa, eu errei aqui.
```suggestion
Para verificar se duas palavras em inglês são homófonas, você pode usar o Dicionário CMU de Pronunciação. E também pode baixá-lo em  http://www.speech.cs.cmu.edu/cgi-bin/cmudict.
```

----

=== Consulta Inversa

Dado um dicionário +d+ e uma chave +k+, é fácil achar o valor correspondente +v = d[k]+. Esta operação é chamada de _consulta_.
(((consulta)))

> Mas e se você tem +v+ e quer achar +k+? Você tem dois problemas: primeiro, pode haver mais de uma chave que mapeia para o valor +v+. Dependendo do que é pedido, você poderia escolher um, ou teria que criar um array que contém todos eles. Segundo, não há uma sintaxe simples que faz uma _consulta invertida_; você tem que procurar.

Reviewer: ```suggestion
Mas e se você tem +v+ e quer achar +k+? Você tem dois problemas: primeiro, pode haver mais de uma chave que mapeia para o valor +v+. Dependendo do que é pedido, você poderia escolher um, ou teria que criar um array que contém todos eles. Segundo, não há uma sintaxe simples que faz uma _consulta inversa_; você tem que procurar.
```

end

> ----

Reviewer: Talvez traduzir invertdict para inverte_dict ou invertedict? Traduzir outros termos dentro da function.

> Cada vez que o loop é percorrido, +key+ recebe uma chave de +d+ e +val+ recebe o valor correspondente. Se +vaç+ não está em +inverse+, isto significa que não a vimos ainda, então criamos um novo item e inicializamos com um _singleton_ (um array que contém um único elemento). Caso contrário nós já vimos esse valor, então nós acrescentamos a chave correspondente ao array.

Reviewer: ```suggestion
Cada vez que o loop é percorrido, +key+ recebe uma chave de +d+ e +val+ recebe o valor correspondente. Se +val+ não está em +inverse+, isto significa que não a vimos ainda, então criamos um novo item e inicializamos com um _singleton_ (um array que contém um único elemento). Caso contrário esse valor já foi visto, e então acrescentamos a chave correspondente ao array.
```

(((singleton)))

Aqui está um exemplo:

[source,@julia-repl-test chap11]
----
julia> hist = histograma("papagaio");

> julia> inverse = invertdict(hist)

Reviewer: Faltou tradução aqui.

Dict{Any,Any} with 2 entries:
  3 => ['a']
  2 => ['p']
  1 => ['g', 'i', 'o']
----

[[fig11-1]]
.Diagrama de Estado
image::images/fig111.svg[]

> <<fig11-1>> é um diagrama de estado mostrando +hist+ e +inverse+. Um dicionário é representado como uma caixa com os pares chave-valor dentro. Se os valores são inteiros, pontos flutuantes ou strings, eu os desenho dentro da caixa, mas eu normalmente desenho arrays fora da caixa, só para simplificar o diagrama.

Reviewer: Sugestão:
```suggestion
<<fig11-1>> é um diagrama de estado mostrando +hist+ e +inverse+. Um dicionário é representado como uma caixa com os pares chave-valor dentro. Para os valores que são inteiros, pontos flutuantes ou strings, eu os desenho dentro da caixa, já para os arrays normalmente desenho fora da caixa, só para simplificar o diagrama.
```

(((diagrama de estado)))

[NOTE]
====

> Eu mencionei anteriormente que um dicionário é implementado usando uma tabela hash e isso significa que as chaves devem ser _hashable_.

Reviewer: Como não tem tradução para hashable, dei essa sugestão:
```suggestion
Eu mencionei anteriormente que um dicionário é implementado usando uma tabela hash e isso significa que as chaves devem ser _hashable_, ou de um tipo que tem uma função hash.
```

(((hashable)))

Uma _hash_ é uma função que recebe um valor (de qualquer tipo) e retorna um inteiro. Dicionários usam estes inteiros, chamados de valores hash, para guardar e consultar pares chave-valor.
(((hash)))
====

[[memos]]
=== Memos

> Se você já brincou com a função +fibonacci+ de <<one_more_exemplo>>, você pode ter percebido que quanto maior o argumento que você fornece, mais tempo a função leva para executar. Além disso, o tempo de execução cresce rapidamente.

Reviewer: ```suggestion
Se você já brincou com a função +fibonacci+ de <<one_more_exemple>>, você pode ter percebido que quanto maior o argumento que você fornece, mais tempo a função leva para executar. Além disso, o tempo de execução cresce rapidamente.
```

> (((fibonnaci)))

Reviewer: @abelsiqueira , essa referência está errada. O correto é 
```suggestion
(((fibonacci)))
```
@abelsiqueira, tem q criar um issue.

> Para entender o por que, considere <<fig11-2>>, que mostra um _grafo de chamada_ para +fibonacci+ com +n = 4+:

Reviewer: ```suggestion
Para entender o porquê, considere <<fig11-2>>, que mostra um _grafo de chamada_ para +fibonacci+ com +n = 4+:
```

(((grafo de chamada)))((("diagram", "grafo de chamada", see="grafo de chamada")))

[[fig11-2]]
.Grafo de chamada
image::images/fig112.svg[]

> Um grafo de chamada mostra um conjunto de quadros de função, com linhas conectando cada quadro à quadros no qual a função chama. No topo do grafo, +fibonacci+ com +n = 4+ chama +fibonacci+ com +n = 3+ e +n = 2+. Por sua vez, +fibonacci+ com +n = 3+ chama +fibonacci+ com +n = 2+ e +n = 1+. E assim em diante.

Reviewer: ```suggestion
Um grafo de chamada mostra um conjunto de quadros da função, com linhas conectando cada quadro aos quadros (que/os quais) a função chama. No topo do grafo, +fibonacci+ com +n = 4+ chama +fibonacci+ com +n = 3+ e +n = 2+. Por sua vez, +fibonacci+ com +n = 3+ chama +fibonacci+ com +n = 2+ e +n = 1+. E assim em diante.
```
Você pode usar ou "que" ou "os quais" no final da 1a frase. Pensa assim: (1) _A função chama os quadros_ ou (2)_A função chama nos quadros_? Se a opção (2) for a mais apropriada, vc usa "no que" ou suas variações como "no qual/nos quais".

> Conte quantas vezes +fibonacci(0)+ e +fibonacci(1)+ são chamadas. Está é uma solução ineficiente do problema, e fica pior a medida em que o argumento aumenta.

Reviewer: ```suggestion
Conte quantas vezes +fibonacci(0)+ e +fibonacci(1)+ são chamadas. Está é uma solução ineficiente do problema, e fica pior à medida em que o argumento aumenta.
```

> Uma solução é acompanhar os valores que já foram calculados armazenando eles em um dicionário. Um valor previamente calculado que é armazenado para uso posterior é chamado de _memo_. Aqui está uma versão “memoizada” de fibonacci:

Reviewer: ```suggestion
Uma solução é acompanhar os valores já calculados armazenando-os em um dicionário. Um valor previamente calculado que é armazenado para uso posterior é chamado de _memo_. Aqui está uma versão “memoizada” de fibonacci:
```

(((memo)))

[source,@julia-setup]
----
conhecidos = Dict(0=>0, 1=>1)

function fibonacci(n)
    if n ∈ keys(conhecidos)
        return conhecidos[n]
    end
    res = fibonacci(n-1) + fibonacci(n-2)
    conhecidos[n] = res
    res
end
----

> +conhecido+ é um dicionáro que acompanha os números de fibonacci que já sabemos. Ele começa com dois itens: +0+ mapeia para +0+ e +1+ mapeia para +1+.

Reviewer: ```suggestion
+conhecido+ é um dicionáro que guarda os números de Fibonacci que já sabemos. Ele começa com dois itens: +0+ mapeia para +0+ e +1+ mapeia para +1+.
```

> Toda vez que +fibonacci+ é chamada, ela checa +conhecidos+. Se o resultado já está lá, ela retorna imediatamente. Por outro lado ela tem que computar um novo valor, adicionar ao dicionário, e retorná-lo.

Reviewer: Sugestão para melhorar a sintaxe da frase.
```suggestion
Toda vez que +fibonacci+ é chamada, ela checa +conhecidos+. Se o resultado já está lá, ela retorna imediatamente. Por outro lado, ela tem que computar um novo valor, que é adicionado no dicionário e devolvido.
```

> Se você executar esta versão de +fibonacci+ e comparar com a original, você verá que ela é muito mais rápida.

Reviewer: ```suggestion
Se você executar esta versão de +fibonacci+ e comparar com a original, você verá que a atual é muito mais rápida.
```

[[global_variables]]
=== Variáveis Globais

> No exemplo anterior, conhecidos é criado fora da função, então ela pertence ao quadro especial chamado +Main+. Variáveis em +Main+ são as vezes chamadas de _globais_ pois podem ser acessadas de qualquer função. Diferente de variáveis locais, que desaparecem quando a função acaba, variáveis globais persistem de uma chamada de função para a próxima.

Reviewer: ```suggestion
No exemplo anterior, conhecidos é criado fora da função, então ela pertence ao quadro especial chamado +Main+. Variáveis em +Main+ são às vezes chamadas de _globais_ pois podem ser acessadas de qualquer função. Diferente de variáveis locais, que desaparecem quando a função acaba, variáveis globais persistem de uma chamada de função para a próxima.
```

> (((global)))((("palavra-chave", "global", see="global")))(((global variable)))

Reviewer: ```suggestion
(((global)))((("palavra-chave", "global", see="global")))(((variável global)))
```

> É comum usar variáveis globais para _flags_; isto é, variáveis booleanas que indicam (“flag”) se uma condição é verdadeira. Por exemplo, alguns programas usam uma flag chamada +verbose+ para controlar o nível de detalhe na saída:

Reviewer: ```suggestion
É comum usar variáveis globais para _flags_; isto é, variáveis booleanas que indicam (“sinal”) se uma condição é verdadeira. Por exemplo, alguns programas usam uma flag chamada +verbose+ para controlar o nível de detalhamento na saída:
```

(((flag)))

[source,@julia-setup]
----
verbose = true

function exemplo1()
    if verbose
        println("Executando exemplo1")
    end
end
----

> Se você tentar re-atribuir uma variável global, você poderá se surpreender. O exemplo a seguir deve acompanhar se a função foi chamada:

Reviewer: ```suggestion
Se tentar reatribuir uma variável global, você poderá se surpreender. O exemplo a seguir deve acompanhar se a função foi chamada:
```

[source,@julia-setup]
----
foi_chamada = false

function exemplo2()
    foi_chamada = true         # ERRADO
end
----

> Mas se você executar ela você verá que o valor de +foi_chamada+ não muda. O problema é que +exemplo2+ cria uma nova variável local chamada +foi_chamada+. A variável local vai embora quando a função termina, e não tem nenhum efeito sobre a variável global.

Reviewer: ```suggestion
Mas se você executar a função, você verá que o valor de +foi_chamada+ não muda. O problema é que +exemplo2+ cria uma nova variável local denominada +foi_chamada+. A variável local é removida quando a função termina, e não tem nenhum efeito sobre a variável global.
```

> Para re-atribuir uma variável global dentro de uma função você deve _declarar_ a variável global antes de usá-la:

Reviewer: ```suggestion
Para reatribuir uma variável global dentro de uma função, você deve _declarar_ a variável global antes de usá-la:
```

> (((declaração)))(((re-atribuição)))

Reviewer: ```suggestion
(((declaração)))(((reatribuição)))
```
Pelo o que eu pesquisei, é reatribuir, td junto.

[source,@julia-setup]
----
foi_chamada = false

function exemplo2()
    global foi_chamada
    foi_chamada = true
end
----

> A _declaração global_ indica ao interpretador algo como, “Nesta função, quando eu digo +foi_chamada+, eu quero dizer a variável global; não crie uma local”

Reviewer: ```suggestion
A _declaração global_ indica ao interpretador algo como “Nesta função, quando eu digo +foi_chamada+, eu quero dizer a variável global; não crie uma local.”
```

(((declaração global)))((("declaração", "global", see="declaração global")))

Aqui está um exemplo que tenta atualizar uma variável global:

[source,@julia-setup chap11]
----
conta = 0

> function exrmplo3()

Reviewer: ```suggestion
function exemplo3()
```

    conta = conta + 1          # ERRADO
end
----

> Se você executá-la, você recebe:

Reviewer: Sugestão para evitar a repetição do "você".
```suggestion
Ao executar a função, você recebe:
```

(((UndefVarError)))

[source,@julia-repl-test chap11]
----
julia> exemplo3()

> ERROR: UndefVarError: count not defined

Reviewer: ```suggestion
ERROR: UndefVarError: conta not defined
```

----

> O Julia assume que +conta+ é local, e a partir desta suposição, de que você está lendo antes de escrevê-la. A solução, novamente, é declarar +conta+ como global.

Reviewer: ```suggestion
O Julia assume que +conta+ é local, partindo da suposição de que você está lendo a função antes de escrevê-la. A solução, novamente, é declarar +conta+ como global.
```

[source,@julia-setup]
----
conta = 0

function exemplo3()
    global conta
    conta += 1
end
----

Se uma variável global se refere a um valor mutável, você pode modificar o valor sem declarar a variável global:

[source,@julia-setup]
----
conhecido = Dict(0=>0, 1=>1)

function exemplo4()
    conhecido[2] = 1
end
----

> Então você pode adicionar, remover e substituir elementos de um array global ou dicionário, mas se você quiser re-atribuir uma variável, você deve declarar ela como global:

Reviewer: ```suggestion
Então você pode adicionar, remover e substituir os elementos de um array global ou dicionário, mas se você quiser reatribuir a variável, você deve declará-la como global:
```

[source,@julia-setup]
----
conhecido = Dict(0=>0, 1=>1)

function exemplo5()
    global conhecido
    conhecido = Dict()
end
----

> Por razões de performance, você deve declarar uma variável global como constante. Você não pode mais re-atribuir a variável mas se ela refere-se a um valor mutável, você pode modificar o valor.

Reviewer: Sugestões para evitar repetição de "você".
```suggestion
Por razões de performance, deve-se declarar uma variável global como constante. Não é mais possível reatribuir a variável mas caso ela faz referência a um valor mutável, você pode modificar o valor.
```

(((const)))((("palavra-chave", "const", see="const")))(((variável global constante)))

[source,@julia-setup]
----
const conhecido = Dict(0=>0, 1=>1)

function exemplo4()
    conhecido[2] = 1
end
----

[WARNING]
====

> Variáveis globais podem ser utéis, mas se você tem muitas delas, e você as modifica frequentemente, elas podem fazer com que programas sejam dificéis de debugar e possuam performance ruim.

Reviewer: ```suggestion
Variáveis globais podem ser utéis, mas se tem muitas delas, e você modifica-as frequentemente, elas podem ser a causa dos programas serem dificéis de debugar e terem performance ruim.
```

====

> === Debugando

Reviewer: Vc acha ruim deixar o próprio termo "Debugging"? Nos capítulos q eu traduzi, em particular esses, eu nunca traduzi o título.

> A medida em que você trabalha com datasets maiores, pode ser que seja difícil debugar imprimindo e checando a saída na mão. Aqui estão algumas sugestões para debugar datasets maiores:

Reviewer: ```suggestion
Na medida em que você trabalha com datasets maiores, pode ser que seja difícil debugar imprimindo e checando a saída na mão. Aqui estão algumas sugestões para debugar datasets maiores:
```
https://www12.senado.leg.br/manualdecomunicacao/redacao-e-estilo/estilo/a-medida-na-medida-em-que

(((debugando)))

* Diminua a entrada:
+

> Se possível, reduza o tamanho do dataset. Por exemplo, se o programa lê um arquivo de texto, comece com apenas as 10 primeiras linhas, ou com o menor exemplo que você encontre que gere um erro. Você não deve editar os arquivos em si, mas modificar o programa para que ele somente leia as primeiras latexmath:[\(n\)] linhas.

Reviewer: ```suggestion
Se possível, reduza o tamanho do dataset. Por exemplo, se o programa lê um arquivo de texto, comece com apenas as 10 primeiras linhas, ou com o menor exemplo que dá erro. Você não deve editar os arquivos em si, mas modificar o programa para que ele leia somente as primeiras latexmath:[\(n\)] linhas.
```

(((redução de tamanho)))
+

> Se existe um erro, você pode reduzir latexmath:[\(n\)] para o menor valor que gera um erro, e em seguida incrementá-lo gradualmente a medida em que você encontra e corrige erros.

Reviewer: ```suggestion
Se existe um erro, você pode reduzir de latexmath:[\(n\)] para o menor valor que dá erro, e em seguida incrementá-lo gradualmente à medida que você encontra e corrige os erros.
```

> * Verifique os resumos e tipos:

Reviewer: ```suggestion
* Verifique os resumos e os tipos:
```

+
Ao invés de imprimir e checar o dataset inteiro, considere imprimir os resumos dos dados: por exemplo, o número de itens em um dicionário ou o total de um array de números.

> (((verificar)))

Reviewer: ```suggestion
(((verificação)))
```

+

> Uma causa comum de erros de execução é um valor que não é do tipo correto. Para debugar esse tipo de erro, é geralmente suficiente imprimir o tipo de um valor.

Reviewer: ```suggestion
Uma causa comum de erros de execução é um valor que não é do tipo correto. Para debugar esse tipo de erro, geralmente é suficiente imprimir o tipo de um valor.
```

* Escreva auto-verificações:
+

> Algumas vezes você pode escrever código para checar erros automaticamente. Por exemplo, se você está calculando a média de um array de números, você poderia checar que o resultado não é maior que o maior elemento do array ou menor que o menor elemento. Isso é chamado de “verificação de sanidade”.

Reviewer: ```suggestion
Algumas vezes você pode escrever código para checar os erros automaticamente. Por exemplo, se você está calculando a média de um array de números, você poderia checar que o resultado não está acima do maior elemento do array ou abaixo do menor elemento. Isso é chamado de “verificação de sanidade”.
```

(((auto-verificação)))(((verificação de sanidade)))
+
Outro tipo de verificação compara o resultado de dois cálculos diferentes para verificar se eles são consistentes. Isso é chamado de “verificação de consistência”.
(((verificação de consistência)))

* Formate a saída:
+

> Resultados de depuração formatados podem facilitar a detecção de erros. Nós vimos um exemplo em <<deb06>>.

Reviewer: ```suggestion
Resultados de debugging formatados podem facilitar a detecção de erros, como visto em um exemplo em <<deb06>>.
```

> (((Formatação)))

Reviewer: ```suggestion
(((formatação)))
```

+

> Novamente, tempo que você gasta construindo andaimes pode reduzir o tempo gasto debugando.

Reviewer: Só sugestões:
```suggestion
E mais uma vez, o tempo que você usa construindo andaimes pode reduzir o tempo gasto no debugging.
```

> (((andaimes)))(((debugando)))

Reviewer: ```suggestion
(((andaime)))(((debugging)))
```

=== Glossário

mapeamento::
Uma relação na qual cada elemento de um conjunto corresponde a um elemento de outro conjunto.
(((mapeamento)))

dicionário::

> Um mapeamento de chaves para seus valores correspondentes.

Reviewer: ```suggestion
Um mapeamento de chaves para os seus valores correspondentes.
```

(((dicionário)))

par chave-valor::
A representação de um mapeamento de uma chave para um valor.
(((par chave-valor)))

item::
Em um dicionário, é outro nome para o par chave-valor.
(((item)))

chave::
Um objeto que aparece em um dicionário como a primeira parte de um par chave-valor.
(((chave)))

valor::
Um objeto que aparece em um dicionário como a segunda parte de um par chave-valor. Isso é mais específico que o nosso uso prévio da palavra “valor”.
(((valor)))

implementação::
Uma maneira de efetuar cálculos.
(((implementação)))

tabela hash::
O algoritmo usado para implementar dicionários em Julia.
(((tabela hash)))

função hash::

> Uma função usado por uma tabela hash para calcular a localização de uma chave.

Reviewer: ```suggestion
Uma função usado por uma tabela hash para determinar a localização de uma chave.
```

Author: Aqui não estaria errado trocar "calcular" por "determinar", mas o algoritmo da tabela hash de fato "calcula" a localização da chave (mas pode ser frescura minha !).

Reviewer: ```suggestion
Uma função usado por uma tabela hash para computar a localização de uma chave.
```
O que vc acha dessa sugestão? Segundo o dicionário, computar é sinônimo de calcular e na minha opinião, me parece até melhor q o "determinar".

(((função hash)))

hashable::
Um tipo que tem uma função hash.
(((hashable)))

consulta::

> Uma operação de dicionário que recebe uma chave e encontra o valor correspondente.

Reviewer: Sugestão só para deixar mais bonito. 
```suggestion
Uma operação em um dicionário que recebe uma chave e encontra o valor correspondente.
```

(((consulta)))

> consulta inversa::

Reviewer: Vc tem que decidir se usa "consulta inversa" ou "consulta invertida". Em particular, eu gosto mais de "consulta inversa".

> Uma operação de dicionário que recebe um valor e encontra uma ou mais chaves que o mapeiam.

Reviewer: ```suggestion
Uma operação em um dicionário que recebe um valor e encontra uma ou mais chaves mapeadas para ele.
```

(((consulta inversa)))

singleton::
Um array (ou outra sequência) com um único elemento.
(((singleton)))

grafo de chamada::

> Um diagrama que mostra todo quadro criado durante a execução do programa, com uma flecha que vai de quem chama para quem é chamado.

Reviewer: ```suggestion
Um diagrama que mostra todo quadro criado durante a execução de um programa, com uma flecha que vai de quem chama para quem é chamado.
```

(((grafo de chamada)))

memo::

> Um valor computado guardado para evitar cálculos futuros desnecessários.

Reviewer: Só sugestões:
```suggestion
Um valor já computado e guardado para evitar cálculos futuros desnecessários.
```

(((memo)))

variável global::
Uma variável definida fora da função. Variáveis globais podem ser acessadas de qualquer função.
(((variável global)))

declaração global::

> Uma declaração que declara um nome de variável como global.

Reviewer: Sugestão para evitar declaração e declara.
```suggestion
Uma declaração para tornar um nome de variável como global.
```

(((declaração global)))

flag::

> Uma variável booleana usada para inidicar se uma condição é verdadeira.

Reviewer: ```suggestion
Uma variável booleana usada para indicar se uma condição é verdadeira.
```

(((flag)))

declaração::

> Uma declaração como +global+ que indica ao interpretador algo sobre a variável.

Reviewer: ```suggestion
Uma declaração como +global+ que informa ao interpretador algo sobre a variável.
```

(((declaração)))

variável global constante::

> Uma variável global que não pode ser re-atribuida.

Reviewer: ```suggestion
Uma variável global que não pode ser reatribuída.
```

(((variável global constante)))

=== Exercícios

[[ex11-1]]
===== Exercício 11-2

> Escreva uma função que leia palavras em _words.txt_ e as guarde como chaves em um dicionário. Não importa quais sejam os valores. Em seguida você pode usar o operador +∈+ como uma maneira rápida de verificar se uma string está ou não no dicionário.

Reviewer: ```suggestion
Escreva uma função que leia palavras em _palavras.txt_ e guarde-as como chaves em um dicionário. Não importa quais sejam os valores. Em seguida, você pode usar o operador +∈+ como uma maneira rápida de verificar se uma string está ou não no dicionário.
```

(((in)))

> Se você fez <<ex10-10>>, você pode comparar a velocidade desta implementação com o operador +∈+ em arrays e a busca bissecção.

Reviewer: Sugestão para tirar repetição de "você":
```suggestion
Se você fez <<ex10-10>>, é possível comparar o tempo desta implementação com o operador +∈+ em arrays e na busca em bissecção.
```

[[ex11-2]]
===== Exercício 11-3

> Leia a documentação da função de dicionário +get!+ e a use para escrever uma versão mais concisa de +invertdict+.

Reviewer: ```suggestion
Leia a documentação da função de dicionário +get!+ e use-a para escrever uma versão mais concisa de +invertdict+.
```
Além disso, não esqueça de traduzir "invertdict".

(((get!)))((("função", "Base", "get!", see="get!")))

[[ex11-3]]
===== Exercício 11-4

> Memoize a função Ackermann de <<ex06-2>> e verifique se a memoização faz possível avaliar a função com argumentos maiores.

Reviewer: ```suggestion
Memoize a função de Ackermann de <<ex06-2>> e verifique se a memoização possibilita a avaliação da função com argumentos maiores.
```

> (((memo)))(((Ackermann function)))

Reviewer: ```suggestion
(((memo)))(((função de Ackermann)))
```

[[ex11-4]]
===== Exercício 11-5

> Se você fez <<ex10-7>>, você já possui uma função chamada +hasduplicates+ que recebe um array como parâmetro e retorna +true+ se há qualquer objeto que aparece mais de uma vez em um array.

Reviewer: Sugestão para tirar repetição de "você":
```suggestion
Se você fez <<ex10-7>>, então já possui uma função chamada +tem_duplas+ que recebe um array como parâmetro e retorna +true+ se há qualquer objeto que aparece mais de uma vez no array.
```

> (((hasduplicates)))

Reviewer: ```suggestion
(((tem_duplas)))
```

> Use um dicionário para escrever uma versão mais rápida, e mais simples de +hasduplicates+.

Reviewer: ```suggestion
Use um dicionário para escrever uma versão mais rápida e simplificada de +tem_duplas+.
```

[[ex11-5]]
===== Exercício 11-6

Duas palavras são “pares rotacionados” se você pode rotacionar um deles e obter o outro (ver +rotacionapalavra+ em <<ex08-5>>).
(((rotacionapalavra)))

> Escreva um programa que lê um array e ache todos os pares rotacionados.

Reviewer: ```suggestion
Escreva um programa que lê um array e encontra todos os pares rotacionados.
```

[[ex11-6]]
===== Exercício 11-7

Aqui está outro quebra cabeça de Car Talk (https://www.cartalk.com/puzzler/browse):
(((Car Talk)))

[quote]
____

> This was sent in by a fellow named Dan O’Leary. He came upon a common one-syllable, five-letter word recently that has the following unique property. When you remove the first letter, the remaining letters form a homophone of the original word, that is a word that sounds exactly the same. Replace the first letter, that is, put it back and remove the second letter and the result is yet another homophone of the original word. And the question is, what’s the word?

Reviewer: João, tem q traduzir aqui tb... Se vc preferir, vc traduz assim:
"Isto foi enviado por uma pessoa chamado Dan O'Leary. Ele encontrou recentemente uma palavra comum de inglês de uma sílaba e cinco letras que tem a seguinte propriedade única. ....."

> Now I’m going to give you an exemplo that doesn’t work. Let’s look at the five-letter word, ‘wrack.’ W-R-A-C-K, you know like to ‘wrack with pain.’ If I remove the first letter, I am left with a four-letter word, ’R-A-C-K.’ As in, ‘Holy cow, did you see the rack on that buck! It must have been a nine-pointer!’ It’s a perfect homophone. If you put the ‘w’ back, and remove the ‘r,’ instead, you’re left with the word, ‘wack,’ which is a real word, it’s just not a homophone of the other two words.

Reviewer: Aqui tb tem q traduzir.

> But there is, however, at least one word that Dan and we know of, which will yield two homophones if you remove either of the first two letters to make two, new four-letter words. The question is, what’s the word?

Reviewer: Aqui tb tem q traduzir.

____

> Você pode usar um dicionário de <<ex11-1>> para verificar se uma string está no array de palavras.

Reviewer: ```suggestion
Você pode usar o dicionário de <<ex11-1>> para verificar se uma string está no array de palavras.
```

[TIP]
====

> Para verificar se duas palavras são homofones, você pode usar o Dicionário CMU de Pronunciação. Você pode baixá-lo em  http://www.speech.cs.cmu.edu/cgi-bin/cmudict.

Reviewer: ```suggestion
Para verificar se duas palavras em inglês são homônimas, você pode usar o Dicionário CMU de Pronunciação. Pode-se baixá-lo em  http://www.speech.cs.cmu.edu/cgi-bin/cmudict.
```

image::images/fig111.svg[]

> <<fig11-1>> é um diagrama de estado mostrando +hist+ e +inverse+. Um dicionário é representado como uma caixa com os pares chave-valor dentro. Para os valores que são inteiros, pontos flutuantes ou strings, eu os desenho dentro da caixa, já para os arrays normalmente desenho fora da caixa, só para simplificar o diagrama.

Reviewer: ```suggestion
<<fig11-1>> é um diagrama de estado mostrando +hist+ e +inverso+. Um dicionário é representado como uma caixa com os pares chave-valor dentro. Para os valores que são inteiros, pontos flutuantes ou strings, eu os desenho dentro da caixa, já para os arrays normalmente desenho fora da caixa, só para simplificar o diagrama.
```

[[memos]]
=== Memos

Se você já brincou com a função +fibonacci+ de <<one_more_exemple>>, você pode ter percebido que quanto maior o argumento que você fornece, mais tempo a função leva para executar. Além disso, o tempo de execução cresce rapidamente.
(((fibonacci)))

Para entender o porquê, considere <<fig11-2>>, que mostra um _grafo de chamada_ para +fibonacci+ com +n = 4+:
(((grafo de chamada)))((("diagram", "grafo de chamada", see="grafo de chamada")))

[[fig11-2]]
.Grafo de chamada
image::images/fig112.svg[]

Um grafo de chamada mostra um conjunto de quadros da função, com linhas conectando cada quadro aos quadros que a função chama. No topo do grafo, +fibonacci+ com +n = 4+ chama +fibonacci+ com +n = 3+ e +n = 2+. Por sua vez, +fibonacci+ com +n = 3+ chama +fibonacci+ com +n = 2+ e +n = 1+. E assim em diante.

> Conte quantas vezes +fibonacci(0)+ e +fibonacci(1)+ são chamadas. Está é uma solução ineficiente do problema, e fica pior à medida em que o argumento aumenta.

Reviewer: Eu deixei passar esse erro na 1a vez.
```suggestion
Conte quantas vezes +fibonacci(0)+ e +fibonacci(1)+ são chamadas. Está é uma solução ineficiente do problema, e fica pior à medida que o argumento aumenta.
```

[[ex11-1]]
===== Exercício 11-2

Escreva uma função que leia palavras em _palavras.txt_ e guarde-as como chaves em um dicionário. Não importa quais sejam os valores. Em seguida, você pode usar o operador +∈+ como uma maneira rápida de verificar se uma string está ou não no dicionário.
(((in)))

> Se você fez <<ex10-10>>, você pode comparar a velocidade desta implementação com o operador +∈+ em arrays e na busca em bissecção.

Reviewer: ```suggestion
Se <<ex10-10>> foi feito, você pode comparar a velocidade desta implementação com o operador +∈+ em arrays e na busca em bissecção.
```

Now I’m going to give you an example that doesn’t work. Let’s look at the five-letter word, ‘wrack.’ W-R-A-C-K, you know like to ‘wrack with pain.’ If I remove the first letter, I am left with a four-letter word, ’R-A-C-K.’ As in, ‘Holy cow, did you see the rack on that buck! It must have been a nine-pointer!’ It’s a perfect homophone. If you put the ‘w’ back, and remove the ‘r,’ instead, you’re left with the word, ‘wack,’ which is a real word, it’s just not a homophone of the other two words.

But there is, however, at least one word that Dan and we know of, which will yield two homophones if you remove either of the first two letters to make two, new four-letter words. The question is, what’s the word?
____

Você pode usar o dicionário de <<ex11-1>> para verificar se uma string está no array de palavras.

[TIP]
====

> Para verificar se duas palavras em inglês são homônimas, você pode usar o Dicionário CMU de Pronunciação. Pode-se baixá-lo em  http://www.speech.cs.cmu.edu/cgi-bin/cmudict.

Reviewer: ```suggestion
Para verificar se duas palavras em inglês são homônimas, você pode usar o Dicionário CMU de Pronunciação. E também pode baixá-lo em  http://www.speech.cs.cmu.edu/cgi-bin/cmudict.
```
